### ISSUE TITLE
Complex command invocation structures

### ISSUE THREAD

**Issue by andrus** [bootique/bootique#131]
This requires more analysis and discussion. Our current "single command per app" design works great - it is easy to understand, command contributions work automatically, etc. However there are often questions from the field on how to do extra things in addition to the main command. A few scenarios:

* I'd like to run Jetty, but also start job scheduler in the same app (so there are 2 **parallel commands** involved).
* I'd like to run healthchecks before the main command can execute (so there are 2 **sequential commands**).

And of course there can be other invocation trees of any complexity. Currently this task can be solved using custom commands that implement the right sequence with a thread pool and CompletableFutures. 

Wondering if we can generalize it though? It may be useful in situations when e.g. you want to run healthchecks and when those pass, start Jetty, all by just combining a bunch of modules, without the need to deal with completable futures.

Also note that command invocation tree is very similar to **job groups** recently introduced in `bootique-job`, where we have a YAML DSL for defining job execution tree. Perhaps commands and jobs are the same thing after all?

**Comment by andrus** [comment]
Here is an example of parallel commands implementation running on the JVM fork-join pool that I am using in some of my apps:
```java
 @Override
public CommandOutcome run(Cli cli) {
    Future<CommandOutcome> o1 = ForkJoinPool.commonPool().submit(() -> commandProvider1.get().run(cli));
    Future<CommandOutcome> o2 = ForkJoinPool.commonPool().submit(() -> commandProvider2.get().run(cli));

    return Stream.of(o1, o2)
            .map(this::extract)
            .reduce(CommandOutcome.succeeded(), this::merge);
}
```
The downside of this implementation is blocking. For N parallel commands it may block N threads in the fork-join pool, even if the commands themselves start a daemon (e.g. ServerCommand from Jetty). Need a concept of "daemon commands" for such cases that start whatever they need to start, and then return without waiting for the daemon to be killed.

**Comment by atomashpolskiy** [comment]
I've been thinking about this for some time now, and it looks to me, that despite the similarities between commands and jobs, these two are still mostly different concepts. Therefore, I suggest limiting this proposal to adding a new standard command, which would be analogous to [JobModule's ExecCommand](https://github.com/bootique/bootique-job/blob/master/bootique-job/src/main/java/io/bootique/job/command/ExecCommand.java) and not bothering with invocation trees, and here's why.

A command is currently synonymous with the application itself. It has access to all program arguments and also the whole configuration (both directly, via ConfigurationFactory, and indirectly, via DI-managed artifacts). With invocation trees, it's kind of expected, that some commands might be run more than once. Sharing configuration imposes the following problems:
(a) It would not be possible to provide different parameter values for different instances of the same command; it's easy to see, how convenient it can be to run multiple instances of the same command at once (e.g. send multiple e-mails via `--send-mail` command), but there's little point to run multiple instances _with the same parameter values_, and any imaginable generic override mechanism is inevitably going to be more cumbersome, than invoking the application with different arguments from a script or using a specialized command (e.g. it would be something like `--send-mail-batch` in the e-mail case).
(b) There would be an increased risk of concurrency issues, when several instances of the same command are run in parallel (shared artifacts and services are not always stateless or thread-safe, and as a developer you don't usually expect the command's code - including underlying services - to be executed in more than one thread simultaneously).

So, the majority of use cases is effectively going to be restricted by simple invocations, like in your first message (run a few different commands in some order or in parallel), and more complicated scenarios can be broken down into a combination of invocations, e.g. (A & B) -> (C) for running health-checks AND some other command before starting the Jetty server. The only problem is the need for the user to handle the return code of the preceding invocation and deciding, if the chain should be terminated or continued. Handling those in a script can be tiresome, but it also has the advantage of distinguishing between different return codes, while a generic chain will only be able to perform a binary check, if the code is zero or non-zero (which might be perfectly fine in most cases though).

So, basically, our task is to strike the balance between flexibility, user convenience and performance. The arguments in favor of supporting custom invocation trees are:
- automatic handling of return codes (convenient, but comes at the price of more sophisticated configuration)
- performance/startup time will obviously be better (at least pre-Java 9)

----

If you'd still like to consider custom invocation trees, then I'd like to suggest a template for YML configuration:
```
commands:
  c1: # alias name for this custom command
    type: command
    command: <string> # name of the command to run
    [forkBootique: true|false] # if true, create new Bootique stack for each invocation; false by default
    [config: <resource-uri>] # provide custom config; main config is used by default
    [mergeConfig: true|false] # if true, merge custom config into main and use the result; true by default
    [environment: {...}] # map of additional Bootique variables
    [allowConcurrentInstances: true|false] # if not present, try finding command annotation; false by default
    [arguments: <string>] # if not present, run with no arguments
    [terminateOnError: true|false] # if true, terminate the processing chain, if the return code is not zero; true by default
  g1:
    type: group
    commands: # list of aliases
      - c1
      - ...
    [serial: true|false] # run commands in order or in parallel; false (parallel) by default
```

Declared custom commands/groups will be run as usual from the CLI (and probably should be listed in help).

Not sure, if all of these parameters can be implemented without inflicting major changes in the core. Maybe some of the proposed parameters are less important than others, and we can postpone their implementation.

Some questions/issues:
- what to do with CLI program arguments for such commands? Override/append/disallow?
- would be nice to provide a way to also use real command names instead of aliases in the group definition; this seems problematic though
